pkg/scaler/prober.go

```diff
@@ -349,22 +380,46 @@ func (p *prober) doProbe(msg string, client kubernetes.Interface, pr *probeResul
 	)
 	for i := 0; i < maxRetries; i++ {
 		if _, err = client.Discovery().ServerVersion(); err == nil {
-			klog.V(4).Infof("%s: probe succeeded", msg)
 			break
 		}
-		klog.V(3).Infof("%s: probe failed with error: %s. Will retry...", msg, err)
+		klog.V(5).Infof("%s: probe failed with error: %s. Will retry...", msg, err)
 	}
 
 	if (err == nil && pr.lastError != nil) || (err != nil && pr.lastError == nil) {
 		pr.resultRun = 0
 	}
 
+	// Check if err is unauthorized or is forbidden and if the failure threshold is not reach.
+	// In case its true try to fetch the secret again and refresh the prober
+	if (apierrors.IsForbidden(err) || apierrors.IsUnauthorized(err) || apierrors.IsInvalid(err)) && pr.resultRun < p.failureThreshold {
```

Comment:
`doProbe` takes in a client and is not really responsible for fetching secrets and refreshing clients. That should be the responsibility of the caller. `doProbe` should be kept simple - given a client it does the probe and returns the error. The caller of `doProbe` inspects the error and if it is one of `IsForbidden` or `IsUnauthorized` or `IsInvalid` then it does not increment the `resultRun` (very bad name), refreshes the clients and skips this run of the probe. In the next run of the probe the new clients will be used. If you feel that that delay is too much then you can either retry calling `doProbe` or reduce the time interval between 2 probes.